model/verification/chunkDataPackRequestStatus.go

```diff
@@ -0,0 +1,43 @@
+package verification
+
+import (
+	"time"
+
+	"github.com/onflow/flow-go/model/flow"
+	"github.com/onflow/flow-go/model/flow/filter"
+)
+
+// ChunkRequestStatus is a data struct represents the current status of fetching
+// chunk data pack for the chunk.
+type ChunkRequestStatus struct {
+	*ChunkDataPackRequest
+	Targets     flow.IdentityList
+	LastAttempt time.Time
+	Attempt     int
```

Comment:
Are these fields needed? The reason why I am asking is because the `ChunkRequests` mempool is concurrently modifying them. However, as the values are not intrinsically protected by a mutex, updates made by `ChunkRequests` are generally _not visible_ in other routines. 

In my opinion, we should either 
* make `ChunkRequestStatus` concurrency safe. E.g. add a mutex to `ChunkRequestStatus` and getter methods for `LastAttempt` and  `Attempt` 
* or remove the field.

Reply:
Please note that **ALL** `ChunkRequests` [methods](https://github.com/onflow/flow-go/blob/0b8af37af5fcdb8c913c36be865b5af133308be5/module/mempool/stdmap/chunk_requests.go#L30-L84) invocations are directly routed to the underlying Backend, which is inherently threadsafe. Hence, all operations on `ChunkRequestStatus` data objects within the `ChunkRequests` mempool are threadsafe and protected by a mutex lock with the atomicity and isolation guaranteed.

In very specific, please consider the (e.g., [`IncrementAttempt`](https://github.com/onflow/flow-go/blob/0b8af37af5fcdb8c913c36be865b5af133308be5/module/mempool/stdmap/chunk_requests.go#L61-L70) method of `ChunkRequests` that is the only method dealing with the mentioned `Attempt` field of `ChunkRequestStatus`. The `IncrementAttempt` directly invokes the `Backend.Run`:

```
func (cs *ChunkRequests) IncrementAttempt(chunkID flow.Identifier) bool {
	err := cs.Backend.Run(func(backdata map[flow.Identifier]flow.Entity) error {
		entity, exists := backdata[chunkID]
		if !exists {
			return fmt.Errorf("not exist")
		}
		chunk := chunkRequestStatus(entity)
		chunk.Attempt++
		chunk.LastAttempt = time.Now()
		return nil
	})

	return err == nil
}
```

Under the hood, `Run` executes the input `func` parameter under a mutual exclusion:
```
// Run executes a function giving it exclusive access to the backdata
func (b *Backend) Run(f func(backdata map[flow.Identifier]flow.Entity) error) error {
	b.Lock()
	defer b.Unlock()
	err := f(b.Backdata.entities)
	b.reduce()
	return err
}
```

So any _concurrent_ execution of `IncrementAttempt` is safely _sequentialized_ by the underlying mutex lock of the `Backend` with preserved atomicity and isolation.

[This test](https://github.com/onflow/flow-go/blob/a0f6777225bc1ce5fc6645b6f74a022a0433345d/module/mempool/stdmap/chunk_requests_test.go#L38-L46) on `IncrementAttempt` of the `ChunkRequests` also asserts the atomicity under concurrent invocations. 

💡  All other methods of `ChunkRequests` (similar to `IncrementAttempt`) are backed by the `mutex` lock of the `Backend` and as long as the underlying `Backend` is threadsafe, so the the `ChunkRequests` methods. 

💡  By relying on the underlying `Backend` that is threadsafe and provides atomicity, we saved one additional lock on the `ChunkRequests` hence to reduce CPU time on lock management.

Reply:
I agree with your assertion that **at the moment** all mutations if `ChunkRequestStatus` are mediated trough the `ChunkRequests` mempool, which uses a mutex. 
* Essentially, what you are saying is: I have an intrinsically unsafe software design, but if you look at the global usage pattern, you will notice that it is used safely. 
* What I am suggesting is: instead of leaving it to the application logic to safely use an intrinsically unsafe software design, please change the software design that it is intrinsically safe. 

Lets take a look at your design:
* in one go-routine we do:
   ```golang
   var chunkRequestsMempool ChunkRequests
   ok = requests.IncrementAttempt(status.ID())
   ```
* in second go-routine we do:
   ```golang
   a := status.Attempt
   ```

There is **no guarantee that the second go routine will _ever_ see the increment the first go routine did**. Your answer to this problem is:
* The second go routine should do:
   ```golang
   requests.ByID(status.ID())
   a := status.Attempt
   ```
    Then, we'll go through the mutex of the mempool and thereby make sure that any update to  `status.Attempt` is visible. This is exactly what I mean I mean by an unsafe software desing:
   * the engineer has to understand the interaction between `ChunkRequests` and `ChunkRequestStatus` to safely use it
   * the engineer has to extra code to safely use `ChunkRequestStatus` in a concurrent setting

Experience has shown (many times), that even the engineers who implemented the original version forget those constrains (some months later) and then introduce subtle concurrency bugs. Even worse for engineers who are just trying to use the code. Do you read every line of the Flow-virtual machine to understand if you are safely using it? Probably not (and you shouldn't be). Ideally, software is designed in a way that its APIs are intrinsically safe. 
## Suggestions

There are a few options to make this design safer:
1. let the stored object be constant (i.e. remove the mutable fields `LastAttempt` and `Attempt` is they are not needed) 
2. Prevent access to the mutable fields `LastAttempt` and `Attempt`, outside of `ChunkRequests`:
   * all the exported methods of `ChunkRequests` could work with `ChunkDataPackRequest` (which presumably is immutable?). Specifically, you would use `ChunkRequests.Add(request *verification.ChunkDataPackRequest)`
   * On `ChunkRequests.Add(request *verification.ChunkDataPackRequest)`, you wrap the `request` into `ChunkRequestStatus`
   * `IncrementAttempt`  and retrieving the number of attempts are fully handled by  `ChunkRequests`. The `ChunkRequestStatus` object would never leave the scope of `ChunkRequests`. _Only then_ you can make the argument that the `Backend`'s mutex provides sufficient concurrency guarantees (because the only way to access the fields is through the `Backend`'s mutex)
3. If you want `ChunkRequestStatus` instances  to leave the scope of `ChunkRequests`, the instances need to be intrinsically concurrency safe:
   * make  `LastAttempt` and `Attempt` private
   * introduce getter plus setter methods for  `LastAttempt` and `Attempt` which go through a mutex

Reply:
Implemented the 2nd suggested approach, i.e., preventing access to mutable fields outside of `ChunkRequests` by [wrapping `ChunkDataRequests` with `ChunkDataPackRequest`](https://github.com/onflow/flow-go/blob/f611b559118a73e832a9767de86d3a9b5fecde2c/module/mempool/stdmap/chunk_requests.go#L50-L55) upon storing in `ChunkRequests` mempool and making sure that [`ChunkRequestStatus` never leaves the `ChunkRequests`](https://github.com/onflow/flow-go/blob/f611b559118a73e832a9767de86d3a9b5fecde2c/module/mempool/stdmap/chunk_requests.go#L84-L93). I have also [made `ChunkRequestStatus` _unexported_](https://github.com/onflow/flow-go/blob/f611b559118a73e832a9767de86d3a9b5fecde2c/module/mempool/stdmap/chunk_requests.go#L84-L93), which makes it solely accessible within the mempool implementation. So it is literally an internal data object representation merely within the `ChunkRequests` mempool. 

P.S: this comment had a large surface of changes including tests and their fixtures. The changes span several commits. For convenience, I made the critical changes hyperlinks in the above explanation ☝️ .